sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/ConnectionProvider.scala

```diff
@@ -18,60 +18,45 @@
-  def getAdditionalProperties(): Properties = new Properties()
+import scala.collection.mutable
 
-  /**
-   * Opens connection toward the database.
-   */
-  def getConnection(): Connection
-}
+import org.apache.spark.internal.Logging
+import org.apache.spark.security.SecurityConfigurationLock
+import org.apache.spark.sql.jdbc.JdbcConnectionProvider
+import org.apache.spark.util.Utils
 
 private[jdbc] object ConnectionProvider extends Logging {
-  def create(driver: Driver, options: JDBCOptions): ConnectionProvider = {
-    if (options.keytab == null || options.principal == null) {
-      logDebug("No authentication configuration found, using basic connection provider")
-      new BasicConnectionProvider(driver, options)
-    } else {
-      logDebug("Authentication configuration found, using database specific connection provider")
-      options.driverClass match {
-        case PostgresConnectionProvider.driverClass =>
-          logDebug("Postgres connection provider found")
-          new PostgresConnectionProvider(driver, options)
-
-        case MariaDBConnectionProvider.driverClass =>
-          logDebug("MariaDB connection provider found")
-          new MariaDBConnectionProvider(driver, options)
-
-        case DB2ConnectionProvider.driverClass =>
-          logDebug("DB2 connection provider found")
-          new DB2ConnectionProvider(driver, options)
-
-        case MSSQLConnectionProvider.driverClass =>
-          logDebug("MS SQL connection provider found")
-          new MSSQLConnectionProvider(driver, options)
-
-        case OracleConnectionProvider.driverClass =>
-          logDebug("Oracle connection provider found")
-          new OracleConnectionProvider(driver, options)
-
-        case _ =>
-          throw new IllegalArgumentException(s"Driver ${options.driverClass} does not support " +
-            "Kerberos authentication")
+  private val providers = loadProviders()
+
+  def loadProviders(): Seq[JdbcConnectionProvider] = {
+    val loader = ServiceLoader.load(classOf[JdbcConnectionProvider],
+      Utils.getContextOrSparkClassLoader)
+    val providers = mutable.ArrayBuffer[JdbcConnectionProvider]()
+
+    val iterator = loader.iterator
+    while (iterator.hasNext) {
+      try {
+        val provider = iterator.next
+        logDebug(s"Loaded built in provider: $provider")
+        providers += provider
+      } catch {
+        case t: Throwable =>
+          logError(s"Failed to load built in provider.", t)
       }
     }
+    // Seems duplicate but it's needed for Scala 2.13
+    providers.toSeq
+  }
+
+  def create(driver: Driver, options: Map[String, String]): Connection = {
+    val filteredProviders = providers.filter(_.canHandle(driver, options))
+    require(filteredProviders.size == 1,
+      "JDBC connection initiated but not exactly one connection provider found which can handle " +
+        s"it. Found active providers: ${filteredProviders.mkString(", ")}")
+    SecurityConfigurationLock.synchronized {
```

Comment:
😭 

I believe this synchronization introduces a significant performance bottleneck for applications that rely on being able to establish many connections simultaneously for performance reasons. This change forces concurrency to 1 when establishing database connections for a given JDBC driver and that strikes me as a significant user impacting change.

Can anyone propose a workaround for this? I have an app that makes connections to thousands of databases and I can't upgrade to any version >3.1.x because of this significant bottleneck.

https://issues.apache.org/jira/browse/SPARK-37391

![so-much-blocking](https://user-images.githubusercontent.com/2316989/142897828-26c2384f-7d09-4e4c-9925-690cc86eab05.jpg)

That screenshot only shows a handful of blocked threads, but in total I have 98 threads blocked waiting for the `SecurityConfigurationLock`.

I do not have this issue when running Spark 3.0.1 with no code changes.

Reply:
@gaborgsomogyi is getConnection() not thread-safe?
Or is there any possibility of finer-grained locking on the actual object that getConnection() is called on?
We could move the filteredProviders.head out of the block but doubt that does anything.

Reply:
> is getConnection() not thread-safe?

Simply no.

This is well thought initially even if it introduces bottlenecks. The main problem is that JVM has a single security context. In order to make a connection one MUST modify the security context, otherwise the JDBC connector is not able to provide the security credentials (TGT, keytab or whatever). Simply saying JDBC connectors are able to get credentials from the single security context.

Since multiple connections are made concurrently (sometimes same DB type w/ different credentials) this must be synchronized not to have a malformed security context (we've made load test and added 10+ DBs and corrupted the security context pretty fast w/o sync and it was horror to find out what's going on).
Workload are coming and going, nobody can tell in advance what kind of security context need to be created at the very beginning of JVM creation. If we would pre-bake the security context (not sure how?!) then the following issues would come:
* How to handle 2 same type of databases (for example MariaDB) with 2 different credentials?
* How to handle if a DB login credentials changed over time?

I personally think the first one can't be solved, the second one could be cured w/ all JVM restarts but I think it's just not user friendly.

If somebody has an excellent idea I would like to hear it.

Reply:
What I can imagine is to pre-set the JVM JAAS config for databases from file like this: `java -Djava.security.auth.login.config=jaas.conf`.
And then a new flag can be introduced like: `spark.jdbc.doNotSyncronize` which is default `true`.
That case security credentials MUSTN'T be provided by Spark's JDBC properties but only from JAAS file.
However this would be super risky and for advanced users only. That said I've spent at least a month to debug the JVM security context what's going on...

Reply:
I don't know anything about the underlying problem, but it had crossed my mind that allowing the synchronization to be optional could be one path forward.

@gaborgsomogyi may I trouble you for some references related to
> The main problem is that JVM has a single security context. In order to make a connection one MUST modify the security context, otherwise the JDBC connector is not able to provide the security credentials (TGT, keytab or whatever)

so I can better familiarize myself with the problem that you are describing?

Beyond this particular issue, what you've shared suggests that the concurrency utilized by my app could be causing us to crosswire data which would be a major problem.

I guess I'd also ask, is there more to it than you described? It sounds like I should either have some cross wired data or if that's not the case then there is some missing piece of the puzzle that means the synchronization is not always required.

Thanks in advance!

Reply:
>  allowing the synchronization to be optional could be one path forward.

Yes, considering the following:
* In case of the JVM security context NOT changed then it's just a useless waste to time.
* In case of the JVM security context changed then it's a must!

I think there are multiple clean solutions:
* Right before calling ` getConnection` the provider could be asked whether it's intended to change the security context and based on that either sync or not.
* We can remove the ` SecurityConfigurationLock.synchronized` here and each and every provider can do it's own sync but only if needed

I would vote on the first solution because it would be less error prone and more clear to provider developers.

Reply:
@srowen @HyukjinKwon I would like to help here either with code or review but I'm going to have surgery either this or next week so I'm going to be fully off for quite some time.

If it's urgent maybe @maropu can help since he's reviewed this code.
All in all I would implement something like this:
```
if (provider.wantsToModifySecCtx()) {
  SecurityConfigurationLock.synchronized {
    filteredProviders.head.getConnection(driver, options)
  }
} else {
  filteredProviders.head.getConnection(driver, options)
}
```

Reply:
Unless somebody has a ground breaking idea the secure providers must be synced, we can free up some time when no authentication is in place.

Reply:
I don't think this is very urgent. We would run Spark 3.2.1 but that's blocked by few issues. I would expect a couple of weeks. cc @huaxingao FYI

Reply:
@tdg5 or @maropu would you be able to find some time for this?

Reply:
The proposed solution seems straightforward to me and I'm happy to take a stab at it. Dumb question though, how would I eventually tweak the spark class loader so it would detect a custom `JdbcConnectionProvider` in my library code?

I tried baking such a class into the uberjar containing the driver app, but I didn't seem to be able to get `ConnectionProvider` to notice my clone of `BasicConnectionProvider`. I was guessing this was due to a class loader disagreement.

My spark-submit looks roughly like so:
```
spark-submit \
  --class theThing \
  --master 'local[*]' \
  --driver-memory 2G \
  --driver-class-path theUberjar.jar
  theUberjar.jar
```

Reply:
@tdg5 Thanks for landing a helping hand! For such scenarios I've created a readme [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/jdbc/README.md).